internal/storage/sbom_store.go

```diff
@@ -0,0 +1,525 @@
+//	   return cur, nil, nil
+//	}, cachedExistingObject
+//
+// )
+//
+//nolint:gocognit,funlen // This functions can't be easily split into smaller parts.
+func (s *sbomStore) GuaranteedUpdate(
+	ctx context.Context,
+	key string,
+	destination runtime.Object,
+	ignoreNotFound bool,
+	preconditions *storage.Preconditions,
+	tryUpdate storage.UpdateFunc,
+	_ runtime.Object,
+) error {
+	out, ok := destination.(*v1alpha1.SBOM)
+	if !ok {
+		return storage.NewInvalidObjError(key, fmt.Sprintf("unexpected out object type: %T", destination))
+	}
+
+	name, namespace := extractNameAndNamespace(key)
+	if name == "" || namespace == "" {
+		return storage.NewInternalErrorf("invalid key: %s", key)
+	}
+
+	for {
+		tx, err := s.db.BeginTxx(ctx, nil)
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			if err := tx.Rollback(); !errors.Is(err, sql.ErrTxDone) {
+				log.Printf("failed to rollback transaction: %v", err)
+			}
+		}()
+
+		query, args, err := sq.Select("*").
+			From("sboms").
+			Where(sq.Eq{"name": name, "namespace": namespace}).
+			ToSql()
+		if err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		sbomRecord := &sbomSchema{}
+		err = tx.GetContext(ctx, sbomRecord, query, args...)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				if !ignoreNotFound {
+					return storage.NewKeyNotFoundError(key, 0)
+				}
+
+				*out = v1alpha1.SBOM{}
+
+				return nil
+			}
+			return err
+		}
+
+		currentSBOM := &v1alpha1.SBOM{}
+		err = json.Unmarshal(sbomRecord.Object, currentSBOM)
+		if err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		err = preconditions.Check(key, currentSBOM)
+		if err != nil {
+			return err
+		}
+
+		updatedSBOM, _, err := tryUpdate(currentSBOM, storage.ResponseMeta{})
+		if err != nil {
+			continue
+		}
+
+		version, err := s.Versioner().ObjectResourceVersion(currentSBOM)
+		if err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+		if err := s.Versioner().UpdateObject(updatedSBOM, version+1); err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		bytes, err := json.Marshal(updatedSBOM)
+		if err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		query, args, err = sq.Update("sboms").
+			Set("object", bytes).
+			Where(sq.Eq{"name": name, "namespace": namespace}).
+			ToSql()
+		if err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		_, err = tx.ExecContext(ctx, query, args...)
+		if err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		sbom, ok := updatedSBOM.(*v1alpha1.SBOM)
+		if !ok {
+			return storage.NewInvalidObjError(key, fmt.Sprintf("unexpected updated object type: %T", updatedSBOM))
+		}
+		*out = *sbom
+
+		if err := s.broadcaster.Action(watch.Modified, updatedSBOM); err != nil {
+			return storage.NewInternalError(err.Error())
+		}
+
+		if err := tx.Commit(); err != nil {
+			return storage.NewInternalError(err.Error())
+		}
```

Comment:
Shouldn't we invert the order:

1. commit to db
2. if everything went fine (the record is into the db) -> notify the watchers

Reply:
You are right